default/python/AI/ColonisationAI.py

```diff
@@ -154,6 +154,54 @@ def max_pop_size():
     return max_pop_size()
 
 
+def get_empire_detection(empire_id):
+    # TODO: move to an EspionageAI module
+    # TODO doublecheck typical AI research times for Radar, for below default value
+    empire_detection = 10 if fo.currentTurn() < 40 else 30
+    empire = None
+    if empire_id != ALL_EMPIRES:
+        empire = fo.getEmpire(empire_id)
+    if empire:
+        # TODO: expose Empire.GetMeter to automatically take into account detection specials
+        for techname in sorted(AIDependencies.DETECTION_STRENGTHS, reverse=True):
+            if empire.techResearched(techname):
+                empire_detection = AIDependencies.DETECTION_STRENGTHS[techname]
+                break
+    return empire_detection
+
+
+def colony_detectable_by_empire(planet_id=None, species_name=None, species_tags=None, empire_id=ALL_EMPIRES,
+                                future_stealth_bonus=0):
```

Comment:
As I think probably still shows in the commit history, I had started by just passing species tags, but then added the ability to be more specific.  The order of precedence is the order of appearance in the signature, unless it can't retrieve the designated planet or species, in which case it falls through to the next variable in line.

I agree that a further improvement would be to let it be able to determine the case for the combination of an arbitrary species on an arbitrary planet.  I'll look at adding that into this, in the 3rd PR instead of here in order to avoid excessive conflicts because of the later move into EspionageAI.  Splitting this into 3 PRs like this has I think facilitated the review, but then it makes some of this followup more complicated.

There is no real significant reason that comes to mind to pass tags instead of a species , that is more of an artifact of how this developed, and I think I left it like that to save a lookup, which however is a pretty trivial thing.  I'll probably drop the tags option.

I suppose I'll reverse the priority of species_name and planet_id, so that if both are specified the function will use the planet specials but the species tags from the provided species rather than whatever is on the planet.

Reply:
On further consideration, I'll just pull forward the first couple of the commits from the third PR into this one, so that I can better deal with the various changes discussed here.